OpenMeteo/fetch_OpenMeteo_data.py

```diff
@@ -0,0 +1,120 @@
+from retry_requests import retry
+
+
+class WeatherDataFetcher:
+    BASE_URL = "https://api.open-meteo.com/v1/"  # Base URL for OpenMeteo API
+
+    def __init__(self):
+        # Initialize the WeatherDataFetcher class
+        # Setup the Open-Meteo API client with cache and retry on error
+        cache_session = requests_cache.CachedSession(".cache", expire_after=3600)
+        retry_session = retry(cache_session, retries=5, backoff_factor=0.2)
+        self.openmeteo = openmeteo_requests.Client(session=retry_session)
+
+    def fetch_forecast_data(self, NWP, params):
+        # Fetch weather data from OpenMeteo API for the specified model (NWP) and parameters
+        url = f"https://api.open-meteo.com/v1/{NWP}"  # Construct API URL
+        try:
+            responses = self.openmeteo.weather_api(url, params=params)  # Get weather data
+            return responses[0]  # Return the first response (assuming only one location)
+        except openmeteo_requests.OpenMeteoRequestsError as e:
+            # Handle OpenMeteoRequestsError exceptions
+            if "No data is available for this location" in str(e):
+                print(f"Error: No data available for the location for model '{NWP}'.")
+            else:
+                print(f"Error: {e}")
+            return None
+
+    def fetch_historical_data(self, params):
+        # Fetch historical weather data from OpenMeteo API
+        BASE_URL = "https://archive-api.open-meteo.com/v1/archive"
+        try:
+            responses = self.openmeteo.weather_api(BASE_URL, params=params)
+            return responses[0] if responses else None
+        except ValueError as e:
+            print(f"Error: {e}")
+            return None
+
+    def process_hourly_data(self, response):
+        # Process hourly data from OpenMeteo API response
+        # Extract hourly data from the response
+        hourly = response.Hourly()
+
+        # Extract variables
+        hourly_variables = {
+            "temperature_2m": hourly.Variables(0).ValuesAsNumpy(),
+            "relative_humidity_2m": hourly.Variables(1).ValuesAsNumpy(),
+            "precipitation": hourly.Variables(2).ValuesAsNumpy(),
+            "cloud_cover": hourly.Variables(3).ValuesAsNumpy(),
+        }
+
+        # Extract time information
+        time_range = pd.date_range(
+            start=pd.to_datetime(hourly.Time(), unit="s", utc=True),
+            end=pd.to_datetime(hourly.TimeEnd(), unit="s", utc=True),
+            freq=pd.Timedelta(seconds=hourly.Interval()),
+            inclusive="left",
+        )
+
+        # Create a dictionary for hourly data
+        hourly_data = {"date": time_range}
+
+        # Assign each variable to the corresponding key in the dictionary
+        for variable_name, variable_values in hourly_variables.items():
+            hourly_data[variable_name] = variable_values
+
+        # Create a DataFrame from the dictionary
+        hourly_dataframe = pd.DataFrame(data=hourly_data)
```

Comment:
For this, we want to have the data be returned in an Xarray Dataset, that has coordinates of `latitude`, `longitude`, and `time_utc`, and then the variables and dataarrays in the Dataset.